### ISSUE TITLE
citation postprocessing/formatting

### ISSUE THREAD

**Issue by stevengj** [bramtayl/BibTeX.jl#3]
Although this could be relegated to a separate package, I think that most people who use BibTeX.jl would want the ability to output citations in a reasonable format.

Features that would be good to have:

* Citation formatting in some [standard bibliography style](https://www.sharelatex.com/learn/Bibtex_bibliography_styles).  I tend to like `ieeetr` myself.  (I like the fact that it uses quotes around article titles, in contrast to book titles in italics, which will look better in text/plain format.)
* Output in text/plain or text/markdown (with the latter supporting links, `\emph`, etcetera).
* Substitution of `@string` variables.
* Conversion of escapes like `{\"{u}}` into unicode `ü`.
* Conversion of article titles to sentence case, excluding literal strings quoted with `{...}`.  e.g `title = {A New {FFT} Algorithm: Awesome, Isn't It?}` should become `"A new {FFT} algorithm: Awesome, isn't it?"`
* ... probably other facets of bibtex I'm forgetting.

**Comment by bramtayl** [comment]
Hmm, I think I might need some help with this. I saw the little bib.py file and thought that it would be fun to port to Julia but a lot of these things look much harder. I have included substitution of `@string` variables already as part of parsing.

**Comment by bramtayl** [comment]
I'm gonna be honest bibtex seems kinda dumb. Reading CSL JSON seems much easier, and there's a package already written for it. Zotero can export it. Getting a working formatter based on one of the CSL XML files also seems doable, instead of porting all of citeproc-js.

**Comment by bramtayl** [comment]
I'd be inclined to do chicago author-date following rmarkdown

**Comment by stevengj** [comment]
Personally, I dislike the Chicago style because of the year placement: *Author, title, Journal volume (year): pages* always seemed ridiculous to me.  I can't recall ever seeing this in any journal I've published in.

**Comment by stevengj** [comment]
bibTeX predates JSON by quite a bit, and there is a large body of data in this format.  Inventing a new format and getting people to use it seems a lot harder than just writing a parser/formatter.

**Comment by bramtayl** [comment]
Does [this ](https://github.com/citation-style-language/styles/blob/master/ieee.csl) correspond to ieeetr? I had a look at the bst file and got very confused.

**Comment by stevengj** [comment]
The easiest thing is to format some example citations and see what they look like.  Reading style files manually is confusing and error-prone in any language.  I linked some examples above, and you can find [more examples here](http://sites.stat.psu.edu/~surajit/present/bib.htm).

**Comment by bramtayl** [comment]
If I could go back and redesign bibtex I'd make the input data come in as something like TOML and the input styles come in as templates. But you can't have everything... I'll see what I can do about de-TeXifying first and then work on output based on ieeetr. @jacobcvt12 maybe we could coordinate.

**Comment by tpoisot** [comment]
I agree with @bramtayl that citeproc JSON seems easier (and supports more document types) -- and you can also write in in YAML, which is fairly simple.

Not to mention that it's better than bibtex at UTF8, etc...

**Comment by bramtayl** [comment]
Yeah UTF8 in particular seems like the real kicker.

**Comment by stevengj** [comment]
UTF8 doesn't seem like a big problem for our purposes.   If you're dealing with an existing bibTeX file, it is straightforward to replace e.g. `\"u` with `ü` as in #6.    And if someone is inputting a citation only for use in Julia, BibTeX.jl handles UTF-8.

The low popularity of Citeproc YAML vs bibTeX makes it a hard sell for me.  There are a *lot* of existing bibliographies out there in bibTeX, particularly in science/math/cs (which is the main use-case for citations in docstrings, after all).

**Comment by tpoisot** [comment]
@stevengj Citeproc JSON is used internally by pandoc, and zotero, and Mendeley. I don't think it's fair to call it unpopular. My points in favor are

1. It handles UTF8 (which bibtex does not, although biblatex does)
2. There is a format for rendering (CSL) which is better than what bibtex has
3. It has more entries types than bibtex, and a far richer data model (with affiliations, grants, licenses, etc)

**Comment by stevengj** [comment]
It may be used internally by Pandoc etc, but I've never met or heard of any scientist or mathematician who actually authors their citation database in Citeproc (JSON or YAML), though I'm sure you can dig up an example online.  It seems like will inevitably impose a big learning curve on authors.

(In principle, we could use `pandoc-citeproc --bib2json foo.bib > foo.json`, but most people won't have pandoc installed.  This was the exact problem in JuliaLang/julialang.github.com#421)

Citeproc also doesn't have any formatting language as far as can tell, e.g. no way to add a note with italics to a reference, and no way to use formulas in a title without typing Unicode math chars directly.

(bibtex has bst files to describe rendering.  But in any case the existence of bst or CSL is mostly irrelevant for our purposes, since it seems like there is zero chance we are going to implement a whole schema language for rendering Julia documentation — it would add way too many dependencies.)

**Comment by stevengj** [comment]
Just the prospect of typing in an author list in Citeproc YAML gives me headaches.  Compare
```
 author:
 - family: Johnson
   given: Steven G.
author:
 - family: Keown
   given: Susan E.
author:
 - family: Johnson
   given: Teresa L.
```
with
```
   authors = {Johnson, Steven G. and Keown, Susan E. and Johnson, Teresa L.}
```
and now imagine entering a paper with 15 authors.   With the bibTeX format I pretty much just copy and paste the author list from the paper and replace `, ` with ` and `.

**Comment by tpoisot** [comment]
Do people still type things manually tho? Regardless, I firmly believe that this package should be able to work with citeproc json as well (and that utf8 support should be a major deciding factor).

I'm just gonna leave it at that, in any case.

**Comment by bramtayl** [comment]
I agree that we need a BibTex parser in Julia; it's a format that is very common and people won't want to convert. Reading in JSON and YAML is a piece of cake because there's already packages for that. So it doesn't seem inconceivable to me that once we support BibTeX start to finish it would be too difficult to support new input formats.

**Comment by bramtayl** [comment]
Ok, so here's my plan here:

1) complete parsing of bibtex into the structure specified by the CSL JSON specification. This means breaking up authors and dates, and some renaming.

2) Start porting citeproc-js to Julia. See how far I get.

The only issue here I think is that citeproc-js has a copy-left license. I know that Base is trying to avoid these kind of licenses. Is that an issue here?

**Comment by lucianolorenti** [comment]
Hi. 
I ported the pybtex formatting utilities to julia. The library is [BibTeXFormat.jl](https://github.com/lucianolorenti/BibTeXFormat.jl). Currently it can generate outputs in latex, markdown, html and plain text. Also it has a bst parser an interpreter that produces latext output.

**Comment by bramtayl** [comment]
Cool!

**Comment by bramtayl** [comment]
maybe we should make an organization?

**Comment by lucianolorenti** [comment]
Yes. maybe we should

**Comment by bramtayl** [comment]
I think there was an offer of including a citation formatter in DocStringExtensions and it seems like we've got all the necessary infrastructure.

**Comment by bramtayl** [comment]
The only important missing component is escapes (at least in my initial parser; I think stevengj's LaTEX parser handles them).

**Comment by lucianolorenti** [comment]
I think that it doesn't handle it either.
```
julia> LaTeXString("{\"{u}}")
L"{\"{u}}"
julia> convert(String,LaTeXString("{\"{u}}"))
"{\"{u}}"
```

I was looking at [latexcodec](https://github.com/mcmtroffaes/latexcodec/blob/develop/latexcodec/codec.py) but I'm not sure how to incoporate the conversion functionality to the parser.

**Comment by stevengj** [comment]
My parser handles these escapes:
```
julia> BibTeX.simplify_latex("{\\\"{u}}")
"ü"
```

**Comment by lucianolorenti** [comment]
You are right. I don't know why I didn't see  simplify_latex. Now I am using it.
